src/core/memory.d

```diff
@@ -809,3 +809,106 @@ struct GC
         gc_runFinalizers( segment );
     }
 }
+
+/**
+ * Pure variants of C's memory allocation functions `malloc`, `calloc` and
+ * `realloc`. Purity is achieved via resetting the `errno` to it's value prior
+ * to being called, removing the function's global state mutation.
+ *
+ * See_Also:
+ *     $(LINK2 https://dlang.org/spec/function.html#pure-functions, D's rules for purity),
+ *     which allow for memory allocation under specific circumstances.
+ */
+void* pureMalloc(size_t size) @trusted pure @nogc nothrow
+{
+    const errno = fakePureGetErrno();
+    void* ret = fakePureMalloc(size);
+    if (!ret || errno != 0)
+    {
+        cast(void)fakePureSetErrno(errno);
+    }
+    return ret;
+}
+/// ditto
+void* pureCalloc(size_t nmemb, size_t size) @trusted pure @nogc nothrow
+{
+    const errno = fakePureGetErrno();
+    void* ret = fakePureCalloc(nmemb, size);
+    if (!ret || errno != 0)
+    {
+        cast(void)fakePureSetErrno(errno);
+    }
+    return ret;
+}
+/// ditto
+void* pureRealloc(void* ptr, size_t size) pure @nogc nothrow
+{
+    const errno = fakePureGetErrno();
+    void* ret = fakePureRealloc(ptr, size);
+    if (!ret || errno != 0)
+    {
+        cast(void)fakePureSetErrno(errno);
+    }
+    return ret;
+}
+
+///
+nothrow @nogc unittest
+{
+    import core.stdc.stdlib : free;
+
+    ubyte[] fun(size_t n)
+    {
+        void* p = pureMalloc(n);
+        assert(n == 0 || p !is null);
+
```

Comment:
I'd inserted `// ...` here to suggest some operations may fail. So please either insert those or (probably better) use `enforce(n == 0 || p !is null)` instead of any assert. That way we throw a legit exception if allocation fails, and the scope statement has a good purpose.